lib/Docstring.ml

```diff
@@ -22,6 +20,11 @@ let parse ~loc text =
   | [] -> Ok (Odoc_parser.ast v)
   | warnings -> Error warnings
 
+let parse_file lx =
+  let location = Lexing.lexeme_start_p lx in
+  let text = Bytes.to_string lx.lex_buffer in
```

Comment:
This is not respecting the API and works only by chance, because the lexbuf is created with `Lexing.from_string`. This stops working however when the buffer is advanced, for example the shbang is skipped in `Parse_with_comments`:

```
#!/usr/bin/truc

{0 Foo}

bar
```

This is unstable because each iterations will duplicate the first line.

The ideal solution would be to modify `odoc-parser`, for example by vendoring it. Otherwise, I see two simpler solutions:

- Changing `Parse.ast` to take `location` and `source` rather than `lexbuf`. The shbang logic could be made by manual string manipulations rather than a lexer.
- Make a "lexbuf to string" function that complies with the API of `Lexing`. This is a bit wasteful as the string will be turned into a lexbuf immediately inside `odoc-parser` but that's the simplest.

Reply:
Out of curiosity would it be difficult to make odoc-parser take a lexbuf instead of string?

I tried the "change the signature" option but couldn't make it work, I will give a shot to the "lexbuf to string" option next time.

Reply:
It wouldn't be difficult at all. odoc-parser is immediately creating a lexbuf out of that string. We still need a workaround until this is done and released (the next release might have a new syntax so we need strong version constraint anyway).

Reply:
done